api/agent/slots.go

```diff
@@ -135,18 +100,23 @@ func (a *slotQueue) ejectSlot(s *slotToken) bool {
 		return false
 	}
 
-	isFound := false
+	foundIdx := -1
 
 	a.cond.L.Lock()
 	for idx, val := range a.slots {
 		if val.id == s.id {
-			a.slots[0], a.slots[idx] = a.slots[idx], a.slots[0]
-			isFound = true
+			foundIdx = idx
 			break
 		}
 	}
-	if isFound {
-		a.slots = a.slots[1:]
+	if foundIdx != -1 {
+		newArr := make([]*slotToken, 0, len(a.slots))
+		for idx, val := range a.slots {
+			if idx != foundIdx {
+				newArr = append(newArr, val)
+			}
+		}
+		a.slots = newArr
```

Comment:
handy page to bookmark https://github.com/golang/go/wiki/SliceTricks

```go
for i := 0; i < len(a.slots); i++ {
  val := a.slots[i]
  if val.id == s.id {
    s.slots = append(s.slots[:i], s.slots[i+1:]...)
    break
  }
}
```

in place append will re-use the underlying array instead of allocating a new one, copy is faster than iteration w/ a branch and one loop (under a lock)